### DISCUSSION TITLE
TP Link Ac 600 on elementary 7.0

### DISCUSSION THREAD

**Discussion by anballes** [elementary/triage#327] [Q&A]
Hi All 

I use elementary OS on Dell Latitude E7440 (its only have 2.4 Ghz wireless card) and bough an new USB wireless card from TP Link, the TP-Link AC600 wireless Realtek RTL8811AU (its 5 Ghz) 

I download the drivers and setup everything, finally the network interface works and can connect to internet.

But the previous process ( im no sure) change something al kernel level then when the system update everything roll back the kernel to the new one and previous setup for TP link dismiss. 

I dont know how setup the wireless card and preserve it in front of new kernel changes.

Anything is welcome and thanks in advance,

Andres

**Comment by desiderantes** [comment]
When you install a driver manually, it relies on the specific kernel you have to generate code to talk to it, and it's susceptible to failures when you update the kernel, so the driver is kept in a place that is specific to that kernel, otherwise the system would just start spewing lots of errors due to the faulty driver. To solve this, it's common to suggest to stick to hardware that contributes their drivers directly to the linux software. In you specific case, there is a package available to add drivers in a way that they get reinstalled automatically when a new kernel is installed (this uses a process called DKMS). After looking for info on your specific card, I found out that if you run in a terminal the command `sudo apt install rtl8812au-dkms` it should, after installation and reboot, allow you to use your usb wifi card with current and future versions, as this will install the driver using DKMS.

**Reply by anballes** [reply]
Hi 

Thanks for prompt answer, before install again the wireless card (its working) ran the command suggest to me but got this error:

Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Errors were encountered while processing:
 rtl8812au-dkms
E: Sub-process /usr/bin/dpkg returned an error code (1)

Any clue ?

**Reply by anballes** [reply]
Another come now, after the above process, the apt upgrade fail 

sudo apt upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Removing old rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Deleting module rtl8812au-4.3.8.12175.20140902+dfsg completely from the DKMS tree.
Loading new rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Errors were encountered while processing:
 rtl8812au-dkms
E: Sub-process /usr/bin/dpkg returned an error code (1)

Thanks for any response

**Reply by desiderantes** [reply]
Oh, the error might be because you haven't removed the driver you manually installed. since it already downloaded the package, it should be OK to do so and try again

**Reply by anballes** [reply]
Thanks Mario, after several reboots the error persist in sudo apt upgrade as well in sudo apt autoremove:

sudo apt upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages will be upgraded:
  libwebp7 libwebpdemux2 libwebpmux3
3 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
3 standard LTS security updates
Need to get 237 kB of archives.
After this operation, 4,096 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Get:1 http://security.ubuntu.com/ubuntu jammy-security/main amd64 libwebp7 amd64 1.2.2-2ubuntu0.22.04.2 [206 kB]
Get:2 http://security.ubuntu.com/ubuntu jammy-security/main amd64 libwebpdemux2 amd64 1.2.2-2ubuntu0.22.04.2 [9,964 B]
Get:3 http://security.ubuntu.com/ubuntu jammy-security/main amd64 libwebpmux3 amd64 1.2.2-2ubuntu0.22.04.2 [20.5 kB]
Fetched 237 kB in 1s (246 kB/s)        
(Reading database ... 159977 files and directories currently installed.)
Preparing to unpack .../libwebp7_1.2.2-2ubuntu0.22.04.2_amd64.deb ...
Unpacking libwebp7:amd64 (1.2.2-2ubuntu0.22.04.2) over (1.2.2-2ubuntu0.22.04.1) ...
Preparing to unpack .../libwebpdemux2_1.2.2-2ubuntu0.22.04.2_amd64.deb ...
Unpacking libwebpdemux2:amd64 (1.2.2-2ubuntu0.22.04.2) over (1.2.2-2ubuntu0.22.04.1) ...
Preparing to unpack .../libwebpmux3_1.2.2-2ubuntu0.22.04.2_amd64.deb ...
Unpacking libwebpmux3:amd64 (1.2.2-2ubuntu0.22.04.2) over (1.2.2-2ubuntu0.22.04.1) ...
Setting up rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Removing old rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Deleting module rtl8812au-4.3.8.12175.20140902+dfsg completely from the DKMS tree.
Loading new rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Setting up libwebp7:amd64 (1.2.2-2ubuntu0.22.04.2) ...
Setting up libwebpmux3:amd64 (1.2.2-2ubuntu0.22.04.2) ...
Setting up libwebpdemux2:amd64 (1.2.2-2ubuntu0.22.04.2) ...
Processing triggers for libc-bin (2.35-0ubuntu3.3) ...
Errors were encountered while processing:
 rtl8812au-dkms
E: Sub-process /usr/bin/dpkg returned an error code (1)

and for autoremove

sudo apt autoremove
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Setting up rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Removing old rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Deleting module rtl8812au-4.3.8.12175.20140902+dfsg completely from the DKMS tree.
Loading new rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Errors were encountered while processing:
 rtl8812au-dkms
E: Sub-process /usr/bin/dpkg returned an error code (1)

My concern if this error will turn in issue for the OS and packages or maybe in some moment wireless card stop working 

Thanks again,

**Reply by desiderantes** [reply]
Ok, so how did you install the driver manually? from where? Did you try to remove it already?

**Reply by anballes** [reply]
Following these instructions 

https://ostechnix.com/install-tp-link-ac600-archer-t2u-nano-wifi-usb-adapter-in-linux/

Without any error during the process

**Reply by desiderantes** [reply]
Then in the same folder you cloned the aircrack repo, run `sudo make dkms_remove`

**Reply by anballes** [reply]
works without any error

sudo make dkms_remove
[sudo] password for andres:             
dkms remove 8812au/5.6.4.2_35491.20191025 --all
Module 8812au-5.6.4.2_35491.20191025 for kernel 6.2.0-32-generic (x86_64).
Before uninstall, this module version was ACTIVE on this kernel.

88XXau.ko:
 - Uninstallation
   - Deleting from: /lib/modules/6.2.0-32-generic/extra/
 - Original module
   - No original module was found for this module on this kernel.
   - Use the dkms install command to reinstall any previous module version.

depmod....
Deleting module 8812au-5.6.4.2_35491.20191025 completely from the DKMS tree.
rm -rf /usr/src/8812au-5.6.4.2_35491.20191025

but after that the sudo apt upgrade and autoremove doesn't work 

sudo apt upgrade
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Calculating upgrade... Done
The following packages have been kept back:
  openssh-client
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Removing old rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Deleting module rtl8812au-4.3.8.12175.20140902+dfsg completely from the DKMS tree.
Loading new rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Errors were encountered while processing:
 rtl8812au-dkms
E: Sub-process /usr/bin/dpkg returned an error code (1)


Any clue ?

**Reply by desiderantes** [reply]
Please run `cat /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log ` and paste the contents here

**Reply by desiderantes** [reply]
If it errors due to permissions, retry with `sudo cat /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log`

**Reply by anballes** [reply]
Thanks, please see

sudo cat /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log
[sudo] password for andres:             
DKMS make.log for rtl8812au-4.3.8.12175.20140902+dfsg for kernel 6.2.0-32-generic (x86_64)
Thu Sep 14 12:25:00 PM -05 2023
make ARCH=x86_64 CROSS_COMPILE= -C /lib/modules/6.2.0-32-generic/build M=/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build  modules
make[1]: Entering directory '/usr/src/linux-headers-6.2.0-32-generic'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: x86_64-linux-gnu-gcc-11 (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
  You are using:           gcc-11 (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_cmd.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.o
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.c: In function ‘aes_cipher’:
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.c:1598:5: warning: this ‘for’ clause does not guard... [-Wmisleading-indentation]
 1598 |     for (j = 0; j < 8; j++)
      |     ^~~
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.c:1601:9: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘for’
 1601 |         payload_index = hdrlen + 8;
      |         ^~~~~~~~~~~~~
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.c: In function ‘aes_decipher’:
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.c:1983:5: warning: this ‘for’ clause does not guard... [-Wmisleading-indentation]
 1983 |     for (j = 0; j < 8; j++)
      |     ^~~
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_security.c:1986:9: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘for’
 1986 |         payload_index = hdrlen + 8;
      |         ^~~~~~~~~~~~~
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_debug.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_io.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_ioctl_query.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_ioctl_set.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_ieee80211.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme.o
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme.c: In function ‘rtw_restructure_ht_ie’:
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme.c:3993:17: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
 3993 |                 if (stbc_rx_enable)
      |                 ^~
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme.c:3996:25: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
 3996 |                         set_mcs_rate_by_mask(ht_capie.supp_mcs_set, MCS_RATE_1R);
      |                         ^~~~~~~~~~~~~~~~~~~~
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme_ext.o
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme_ext.c: In function ‘mgt_dispatcher’:
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme_ext.c:784:27: warning: this statement may fall through [-Wimplicit-fallthrough=]
  784 |                         if(check_fwstate(pmlmepriv, WIFI_AP_STATE) == _TRUE)
      |                           ^
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_mlme_ext.c:789:17: note: here
  789 |                 case WIFI_ASSOCREQ:
      |                 ^~~~
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_wlan_util.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_vht.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_pwrctrl.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_rf.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_recv.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_sta_mgt.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_ap.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_xmit.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_p2p.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_tdls.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_br_ext.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_iol.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_sreset.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_btcoex.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_beamforming.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/rtw_odm.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/core/efuse/rtw_efuse.o
  CC [M]  /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.o
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.c: In function ‘rtw_change_ifname’:
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.c:2183:28: warning: passing argument 1 of ‘_rtw_memcpy’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
 2183 |         _rtw_memcpy(pnetdev->dev_addr, padapter->eeprompriv.mac_addr, ETH_ALEN);
      |                     ~~~~~~~^~~~~~~~~~
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.c:807:24: note: expected ‘void *’ but argument is of type ‘const unsigned char *’
  807 | void _rtw_memcpy(void* dst, const void* src, u32 sz)
      |                  ~~~~~~^~~
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.c: In function ‘rtw_random32’:
/var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.c:2310:16: error: implicit declaration of function ‘prandom_u32’; did you mean ‘get_random_u32’? [-Werror=implicit-function-declaration]
 2310 |         return prandom_u32();
      |                ^~~~~~~~~~~
      |                get_random_u32
cc1: some warnings being treated as errors
make[2]: *** [scripts/Makefile.build:260: /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/os_dep/osdep_service.o] Error 1
make[1]: *** [Makefile:2026: /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-6.2.0-32-generic'
make: *** [Makefile:1457: modules] Error 2

**Reply by desiderantes** [reply]
Yeah, kernel seems to be too new for that, you might have to download the .deb package from [here](https://packages.ubuntu.com/mantic/all/rtl8812au-dkms/download), running `sudo apt remove rtl8812au-dkms`  to remove the one failing, and installing it from the downloaded deb (You can use either the Eddy app available in the App Center, or via the `dpkg` tool in console)

**Reply by anballes** [reply]
Thanks, the sudo apt remove rt18812au-dmks was successful but the package installation from Synaptic got an error:

Selecting previously unselected package rtl8812au-dkms.
(Reading database ... 159778 files and directories currently installed.)
Preparing to unpack .../rtl8812au-dkms_4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1_all.deb ...
Unpacking rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Setting up rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Loading new rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Errors were encountered while processing:
 rtl8812au-dkms
E: Sub-process /usr/bin/dpkg returned an error code (1)
A package failed to install.  Trying to recover:
Setting up rtl8812au-dkms (4.3.8.12175.20140902+dfsg-0ubuntu17~22.04.1) ...
Removing old rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Deleting module rtl8812au-4.3.8.12175.20140902+dfsg completely from the DKMS tree.
Loading new rtl8812au-4.3.8.12175.20140902+dfsg DKMS files...
Building for 6.2.0-32-generic
Building initial module for 6.2.0-32-generic
Error! Bad return status for module build on kernel: 6.2.0-32-generic (x86_64)
Consult /var/lib/dkms/rtl8812au/4.3.8.12175.20140902+dfsg/build/make.log for more information.
dpkg: error processing package rtl8812au-dkms (--configure):
 installed rtl8812au-dkms package post-installation script subprocess returned error exit status 10
Errors were encountered while processing:
 rtl8812au-dkms

**Reply by desiderantes** [reply]
Weird, the link I shared gives you a .deb called `rtl8812au-dkms_4.3.8.12175.20140902+dfsg-0ubuntu20_all.deb` (for the ubuntu release codenamed `mantic`) but the log you shared is for `rtl8812au-dkms_4.3.8.12175.20140902+dfsg-0ubuntu17_all.deb` (for the ubuntu release codenamed `kinetic`) You need the one from mantic because it is the one compatible with the 6.x kernels

**Reply by anballes** [reply]
Hi Mario

I think I will opt to do a fresh install of the OS and boot from scratch, if I go this way what would you suggest me to do a clean install of the drivers and not fill the system with software patches ?

**Comment by anballes** [comment]
Back again guys

I decided that I will do a clean install of elementary 7 and the first thing will be to install the driver card, but please could you confirm me what would be the steps to just installed the O.S, install the card but remain persistent so that with kernel updates does not stop working ?

**Reply by anballes** [reply]
Hi All

Please, any clue ?

Thanks in advance.

**Comment by jeremyb31** [comment]
See https://github.com/aircrack-ng/rtl8812au

**Reply by anballes** [reply]
Hi Jeremy 

Thanks for the contribution.

Unfortunately I got this error during the installation:

sudo make dkms_install
mkdir: created directory '/usr/src/8812au-5.6.4.2_35491.20191025'
cp -r * /usr/src/8812au-5.6.4.2_35491.20191025
dkms add -m 8812au -v 5.6.4.2_35491.20191025
Creating symlink /var/lib/dkms/8812au/5.6.4.2_35491.20191025/source -> /usr/src/8812au-5.6.4.2_35491.20191025
dkms build -m 8812au -v 5.6.4.2_35491.20191025
Error! Your kernel headers for kernel 5.15.0-58-generic cannot be found at /lib/modules/5.15.0-58-generic/build or /lib/modules/5.15.0-58-generic/source.
You can use the --kernelsourcedir option to tell DKMS where it's located.
make: *** [Makefile:1788: dkms_install] Error 1

Later on, try again to run the make with flag suggested but nothing , finally when I ran again the command:

sudo make dkms_install
cp -r * /usr/src/8812au-5.6.4.2_35491.20191025
dkms add -m 8812au -v 5.6.4.2_35491.20191025
Error! DKMS tree already contains: 8812au-5.6.4.2_35491.20191025
You cannot add the same module/version combo more than once.
make: *** [Makefile:1787: dkms_install] Error 3


Any clue ?

**Reply by jeremyb31** [reply]
sudo apt install linux-headers-$(uname -r)

**Reply by anballes** [reply]
Great, you right. 

My problem was that in a fresh install of elementary it does not download the kernel source, so as a step before _s_udo_ make dkms_install_ you would have to verify the kernel and that the OS has git.

Thanks to all
